From c713dc1652a314dcf62f887e5f3b3fb22194e8ac Mon Sep 17 00:00:00 2001
From: Nadav Haklai <nadavh@marvell.com>
Date: Tue, 8 Sep 2015 13:48:17 +0300
Subject: [PATCH 102/150] ARM: mvebu: Optimize L2 cache performance and power
 consumption

Configure L2 Prefetch control register:
* Incr double linefill enable
* Data prefetch enable
* Double linefill enable
* Double linefill on WRAP disable
* NO prefetch drop enable

Configure L2 Power control register:
* Enable Dynamic clock gating

Signed-off-by: Nadav Haklai <nadavh@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/23492
Tested-by: Star_Automation <star@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/23673

Conflicts:
	arch/arm/mach-mvebu/board-v7.c

Change-Id: I5182e9462e1e7cc03b1beb05023b24ce2a4080db
Reviewed-on: http://vgitil04.il.marvell.com:8080/27308
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Tested-by: Lior Amsalem <alior@marvell.com>
[zou:Original patch taken from
https://github.com/MarvellEmbeddedProcessors/linux-marvell-prv.git linux-4.4.1-devel]
Signed-off-by: zou cao <cao.zou@windriver.com>
---
 arch/arm/mach-mvebu/board-v7.c |   35 +++++++++++++++++++++++++++++++++++
 1 files changed, 35 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-mvebu/board-v7.c b/arch/arm/mach-mvebu/board-v7.c
index 44f6cd4..9aa9c28 100644
--- a/arch/arm/mach-mvebu/board-v7.c
+++ b/arch/arm/mach-mvebu/board-v7.c
@@ -114,8 +114,43 @@ static void __init mvebu_memblock_reserve(void)
 static void __init mvebu_memblock_reserve(void) {}
 #endif
 
+void __init mvebu_l2_optimizations(void)
+{
+	void __iomem *l2x0_base;
+	struct device_node *np;
+	unsigned int val;
+
+	np = of_find_compatible_node(NULL, NULL, "arm,pl310-cache");
+	if (!np)
+		return;
+
+	l2x0_base = of_iomap(np, 0);
+	if (!l2x0_base) {
+		of_node_put(np);
+		return;
+	}
+
+	/* Configure the L2 PREFETCH and POWER registers */
+	val = 0x58800000;
+	/*
+	*  Support the following configuration:
+	*  Incr double linefill enable
+	*  Data prefetch enable
+	*  Double linefill enable
+	*  Double linefill on WRAP disable
+	*  NO prefetch drop enable
+	 */
+	writel_relaxed(val, l2x0_base + L310_PREFETCH_CTRL);
+	val = L310_DYNAMIC_CLK_GATING_EN;
+	writel_relaxed(val, l2x0_base + L310_POWER_CTRL);
+
+	iounmap(l2x0_base);
+	of_node_put(np);
+}
+
 static void __init mvebu_init_irq(void)
 {
+	mvebu_l2_optimizations();
 	irqchip_init();
 	mvebu_scu_enable();
 	coherency_init();
-- 
1.7.5.4

